Update expo-cli.md (#18847)
expo/expo: docs/pages/workflow/expo-cli.md

import { Terminal } from '~/ui/components/Snippet';

The `expo` package provides a small and powerful CLI tool `npx expo` which is designed to keep you moving fast during app development.

> **Highlights**

Reviewer: ```suggestion
**Highlights**:
```

## Develop

> Start a development server for working on your project by running:

Reviewer: ```suggestion
Start a development server to work on your project by running:
```

- `r`: Reload the app on any connected device.
- `m`: Open the dev menu on any connected native device (web not supported).
- `shift` + `m`: Choose more commands to trigger on connected devices. This includes toggling the performance monitor, opening the element inspector, reloading the device, and opening the dev menu.
- `j`: Open Chrome Dev Tools for any connected device that is using Hermes as the JavaScript engine. [Learn more](/guides/using-hermes#javascript-inspector-for-hermes).
- `o`: Open project code in your editor. This can be configured with the `EXPO_EDITOR` and `EDITOR` [environment variables](#environment-variables).
- `c`: Show development server URL as a QR code in the terminal.
- `?`: Show all Terminal UI commands.

### Server URL

By default, the project is served over a LAN connection. You can change this behavior to localhost-only by using the flag `npx expo start --localhost`.

Other available options are:

- `--port`: Port to start the dev server on (does not apply to Webpack or [tunnel URLs](#tunneling)). Default: **19000**.
- `--https`: Start the dev server using a secure origin. This is currently only supported on web.

> You can force the URL to be any value with the `EXPO_PACKAGER_PROXY_URL` environment variable, for example:

Reviewer: ```suggestion
You can force the URL to be any value with the `EXPO_PACKAGER_PROXY_URL` environment variable. For example:
```

<Terminal cmd={[
'export EXPO_PACKAGER_PROXY_URL=http://expo.dev',
'$ npx expo start'
]} />

Will open apps to: `exp://expo.dev:80` (the `:80` is a temporary workaround for Android websockets).

#### Tunneling

Restrictive network conditions (common for public Wi-Fi), firewalls (common for Windows users), or Emulator misconfiguration can make it difficult to connect a remote device to your dev server over lan/localhost.

Sometimes it's easier to connect to a dev server over a proxy URL that's accessible from any device with internet access, this is referred to as **tunneling**. `npx expo start` provides built-in support for **tunneling** via [ngrok][ngrok].

> To enable tunneling, first install `@expo/ngrok`:

Reviewer: ```suggestion
To enable tunneling, first install `@expo/ngrok` as global package:
```

Author: It can be installed either globally or locally, in this case I'm just keeping the setup simple and leaning on the installation command.

Reviewer: Okay, wasn't aware of that. I'll remove this suggestion.

#### Compiling iOS

An iOS app can have multiple **schemes** for representing different sub-apps like App Clips, watchOS apps, Safari Extensions, and so on. By default, `expo run:ios` will choose the scheme for your iOS app. You can pick a custom scheme with the `--scheme <my-scheme>` argument. If you pass in the `--scheme` argument alone, then Expo CLI will prompt you to choose a scheme from the list of available options in your Xcode project.

The scheme you select will filter out which `--device` options show up in the selection prompt, for example, selecting an Apple TV scheme will only show available Apple TV devices.

You can compile an iOS app for production by running:

<Terminal cmd={['$ npx expo run:ios --configuration Release']} />

This build is not automatically code signed for submission to the Apple App Store. `npx expo run:ios` should mostly be used to test bugs that only show up in production builds. Native code signing requires several network requests and is prone to many different types of errors from the Apple servers. To generate a production build that is code signed for the App Store, we recommend using [EAS Build](/build/introduction).

> When you compile your app onto a Simulator, the Simulator's native error logs will be piped to the Expo CLI process in the terminal. This is useful for quickly seeing bugs that may cause a fatal error, for example missing permission messages. Error piping is not available for physical iOS devices.

Reviewer: ```suggestion
When you compile your app onto a Simulator, the Simulator's native error logs will be piped to the Expo CLI process in the terminal. This is useful for quickly seeing bugs that may cause a fatal error. For example, missing permission messages. Error piping is not available for physical iOS devices.
```

<Terminal cmd={['$ open -a /Applications/Android\ Studio.app android']} />

### Exporting

You can export the JavaScript and assets for your app using Metro bundler by running:

<Terminal cmd={[
'$ npx expo export',
]} />

This is done automatically when using `eas update` or when compiling the native runtime. The `export` command works similar to most web frameworks:

- A bundler transpiles and bundles your application code for **production** environments, stripping all code guarded by the `__DEV__` boolean.
- All static files are copied into a static `dist/` folder which can be served from a static host.
- Contents of the `public/` folder are copied into the `dist/` folder as-is.

The following options are provided:

- `--platform <platform>`: Choose the platform to compile for: 'ios', 'android', 'all'. **Default: all**. 'web' is also available if configured in the Expo config. For more information, see [Customizing Metro](/guides/customizing-metro).
- `--dev`: Bundle for **development** environments without minifying code or stripping the `__DEV__` boolean.
- `--output-dir <dir>`: The directory to export the static files to. **Default: dist**

> - `--max-workers <number>`: Maximum number of tasks to allow the bundler to spawn. Setting this to `0` will run all transpilation on the same process, meaning you can easily debug Babel transpilation.

Reviewer: ```suggestion
- `--max-workers <number>`: Maximum number of tasks to allow the bundler to spawn. Setting this to `0` will run all transpilation on the same process. This means you can easily debug Babel transpilation.
```

'$ npx expo prebuild',
]} cmdCopy="npx expo prebuild" />

Native source code must be generated before a native app can compile. Expo CLI provides a unique and powerful system called _prebuild_, that generates the native code for your project. To learn more, read the [Expo Prebuild docs](/workflow/prebuild.md).

## Config

Evaluate the Expo config (**app.json**, or **app.config.js**) by running:

<Terminal cmd={[
'$ npx expo config',
]} cmdCopy="npx expo config" />

> - `--full`: Include all project config data

Reviewer: ```suggestion
- `--full`: Include all project config data.
```

The other two bullet points are ending with a period. For consistency, I'm adding one here.

- `--json`: Output in JSON format, useful for converting an `app.config.js` to an `app.config.json`.
- `-t, --type`: [Type of config](#config-type) to show.

### Config Type

There are three different config types that are generated from the Expo config:

> - `public`: The manifest file for use with OTA updates, think of this like an `index.html` file's `<head />` element but for native apps.

Reviewer: ```suggestion
- `public`: The manifest file to use with OTA updates. Think of this like an `index.html` file's `<head />` element but for native apps.
```

- `prebuild`: The config that is used for [Expo Prebuild](/workflow/prebuild) including async modifiers. This is the only time the config is not serializable.
- `introspect`: A subset of the `prebuild` config that only shows in-memory modifications like `Info.plist` or `AndroidManifest.xml` changes. Learn more about [introspection](/guides/config-plugins#introspection).

## Install

> Unlike the web, React Native is not backwards compatible, meaning npm packages often need to be the exact right version for the currently installed copy of `react-native` in your project. Expo CLI provides a best-effort tool for doing this using a list of popular packages and the known working version combinations. Simply use the `install` command as a drop-in replacement for `npm install`:

Reviewer: ```suggestion
Unlike the web, React Native is not backwards compatible.  This means that npm packages often need to be the exact right version for the currently installed copy of `react-native` in your project. Expo CLI provides a best-effort tool for doing this using a list of popular packages and the known working version combinations. Simply use the `install` command as a drop-in replacement for `npm install`:
```

<Terminal cmd={[
'$ npx expo install expo-camera',
]} cmdCopy="npx expo install expo-camera" />

> You can add as many packages as you'd like:

Reviewer: Running a single instance of this command, you can also install multiple packages:

Author: ```suggestion
Running a single instance of this command, you can also install multiple packages:
```

<Terminal cmd={[
'$ npx expo install typescript expo-sms',
]} cmdCopy="npx expo install typescript expo-sms" />

You can directly pass arguments to the underlying package manager by using the `--` operator:

<Terminal cmd={[
'$ yarn expo install typescript -- -D',
'# yarn add typescript -D',

> ]} cmdCopy="yarn expo install typescript -- -D" />

Reviewer: I'm curious to know using `yarn`, do you want to showcase that a different package manager can be used?  I think this is clever to demonstrate different package managers as examples 💯

### Version validation

> You can also perform validation and correction with the `--fix` and `--check` flags:

Reviewer: ```suggestion
You can perform validation and correction with the `--check` and `--fix` flags:
```

The validation relates to `check` correction to `fix` . we don't want folks to get confused 😄

- `--check`: Check which installed packages need to be updated.
- `--fix`: Automatically update any invalid package versions.

Example:

<Terminal cmd={[
'# Check all packages for incorrect versions, prompt to fix locally',
'$ npx expo install --check',
]} cmdCopy="npx expo install --check" />

> `npx expo install --check` prompts you to install incorrect packages locally and exits with non-zero in CI, meaning you can use this to do continuous immutable validation. In contrast `npx expo install --fix` will always fix packages if needed, regardless of the environment.

Reviewer: ```suggestion
`npx expo install --check` prompts you about packages that are installed incorrectly. It also prompts about installing the these packages to their compatible versions locally. It exits with non-zero in Continuous Integration (CI). This means you can use this to do continuous immutable validation. In contrast, `npx expo install --fix` will always fix packages if needed, regardless of the environment.
```
